packages/config/lib/index.ts

```diff
@@ -1,206 +1,101 @@
-    this.ENVIRONMENT.name = PROJECT_PACKAGE.name;
-    this.ENVIRONMENT.version = PROJECT_PACKAGE.version;
-    this.ENVIRONMENT.path = projectPath;
-    // unique instance ID (6 char)
-    this.ENVIRONMENT.nodeId =
-      this.ENVIRONMENT.nodeId ||
-      randomBytes(20)
-        .toString("hex")
-        .substr(5, 6);
-    // wait until core config is set
-    this.ENVIRONMENT.namespace = this.config.Config.namespace;
-
-    // put config into DI
-    Container.set("ENVIRONMENT", this.ENVIRONMENT);
-  }
+    ConfigMergeHelper.checkForMissingConfigProperties(configModule);
 
-  /* HELPER */
-  private deepMapHelper(obj, callback, nestedPath = "") {
-    Object.entries(obj).map((entry) => {
-      const newPath = `${nestedPath}${entry[0]}.`;
-      typeof entry[1] === "object" && entry[1] != null
-        ? this.deepMapHelper(entry[1], callback, newPath)
-        : callback(entry[0], entry[1], newPath.slice(0, -1)); // remove last dot on last round
-    });
+    this.config[name] = configModule;
   }
 
-  /* PUBLIC */
-
-  // register config for module and return this initialized configuration
-  public registerConfig(moduleName: string, moduleConfigPath: string): any {
-    // check if path was already included, otherwise just return result
-    if (this.configModules.find((configModule) => configModule.name === moduleName) == null) {
-      const configModule: IConfigModule = {
-        name: moduleName,
-        path: moduleConfigPath
-      };
-      this.configModules.push(configModule);
-
-      // apply config to global config object
-      return this.applyConfig(moduleName, moduleConfigPath);
-    } else {
-      return this.getConfig(moduleName);
-    }
+  public registerConfig(name: string, configDirectory: string): object {
+    const baseConfigModule = this.getConfigFromConfigFiles(configDirectory);
+    this.applyConfigModule(name, baseConfigModule);
+
+    return this.getConfig(name);
   }
 
-  public getConfig(moduleName?: string): any {
-    const config = Container.get("CONFIG");
+  public registerApplicationConfigModule(name: string): object {
+    const baseConfigModule = {};
+    this.applyConfigModule(name, baseConfigModule);
 
-    if (moduleName == null) {
-      // check if config is in its final state (after boot or during boot)
-      if (this.bootLoader.hasBooted() || this.bootLoader.isBooting()) {
-        throw Error("Configuration not available before booting.");
-      }
+    return this.getConfig(name);
+  }
 
-      // return copy instead of a ref
-      return { ...config };
-    } else {
-      // module configuraion is always final when available
-      // return copy instead of a ref
-      return { ...config[moduleName] };
+  public getConfig(name?: string): any {
+    if (name == null) {
+      if (!this.bootLoader.hasBooted() || this.bootLoader.isBooting()) {
+        throw new Error("Configuration not available before booting.");
+      }
+      return { ...this.config };
```

Comment:
I would rather do a deep clone here instead of a shallow one. eg stringify and parse, or using lodash. preferably lodash.